src/masternodes/validation.cpp

```diff
@@ -2400,7 +2402,182 @@ static Res ValidateCoinbaseXVMOutput(const XVM &xvm, const FinalizeBlockCompleti
     return Res::Ok();
 }
 
-static Res ProcessEVMQueue(const CBlock &block, const CBlockIndex *pindex, CCustomCSView &cache, const CChainParams& chainparams, const uint64_t evmQueueId) {
+CTransferDomainMessage DecodeTransferDomainMessage(const CTransactionRef& tx, const CBlockIndex* pindex, const CChainParams& chainparams) {
+    std::vector<unsigned char> metadata;
+    CustomTxType txType = GuessCustomTxType(*tx, metadata);
+    if (txType == CustomTxType::TransferDomain) {
+        auto txMessage = customTypeToMessage(txType);
+        if (CustomMetadataParse(pindex->nHeight, chainparams.GetConsensus(), metadata, txMessage))
+            return *(std::get_if<CTransferDomainMessage>(&txMessage));
+    }
+    return CTransferDomainMessage{};
+}
+
+CAmount GetEvmDST20TotalSupply(const DCT_ID& id, const std::string &stateRoot = std::string()) {
+    auto result = XResultValue(evm_try_get_dst20_total_supply(result, id.v, stateRoot));
+        if (result)
+            if (auto balance = *result)
+                return static_cast<CAmount>(balance);
+
+    return -1;
+}
+
+CAmount GetEvmDFIBalance(const CScript& address, const std::string &stateRoot = std::string()) {
+    CTxDestination dest;
+    if (ExtractDestination(address, dest) && dest.index() == WitV16KeyEthHashType) {
+        const auto keyID = std::get<WitnessV16EthHash>(dest);
+        auto result = XResultValue(stateRoot.empty() ? evm_try_get_balance(result, keyID.ToHexString()) : evm_try_get_balance_at_state_root(result, keyID.ToHexString(), stateRoot));
+        if (result)
+            if (auto balance = *result)
+                return static_cast<CAmount>(balance);
+    }
+
+    return -1;
+}
+
+void ProcessAccountingStateBeforeBlock(const CBlock &block, const CBlockIndex* pindex,  CCustomCSView &mnview, const CChainParams& chainparams, CEVMInitialState& evmInitialState) {
+    std::map<CScript, CStatsTokenBalances> &evmBalances = evmInitialState.evmBalances;
+    CStatsTokenBalances& dst20EvmTotalSupply = evmInitialState.dst20EvmTotalSupply;
+
+    // Storing inital EVM address balances
+    for (const auto &tx : block.vtx) {
+        auto txMessage = DecodeTransferDomainMessage(tx, pindex, chainparams);
+        for (auto const &[src, dst]: txMessage.transfers){
+            if (src.amount.nTokenId == DCT_ID{0}) {
+                std::vector<std::tuple<const uint8_t, VMDomain, const CScript&, const CTokenAmount&, bool>> balanceList{
+                    { src.domain, VMDomain::EVM, src.address, src.amount, evmBalances.find(src.address) == evmBalances.end() },
```

Comment:
Doesn't seem necessary to track `src.amount` here.

Reply:
Simplified a bit